### PR TITLE
Sql Assessment HTML report

### PR DESCRIPTION
**Pull request by vchvlad**
We need to add some reporting functionality. Thus, we'd like to start with a simple HTML report.

### CODE REVIEW

File: src/sql/workbench/contrib/assessment/browser/asmtResultsView.component.ts [hunk 1]
```diff
@@ -106,17 +109,19 @@ export class AsmtResultsViewComponent extends TabChild implements IAssessmentCom
...
 	private isServerMode: boolean;
 	private rowDetail: RowDetailView<Slick.SlickData>;
 	private exportActionItem: IAction;
+	private generateReportActionItem: IAction;
 	private placeholderElem: HTMLElement;
 	private placeholderNoResultsLabel: string;
 	private spinner: { [mode: number]: HTMLElement } = Object.create(null);
-	private lastInvokedResults: azdata.SqlAssessmentResultItem[];
+	private lastInvokedResult: SqlAssessmentResultInfo;
+	private initialConnectionInfo: any;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/browser/asmtResultsView.component.ts:124]: You don't have a more specific type you can give this? Like ConnectionInfo?

File: src/sql/workbench/contrib/assessment/browser/asmtResultsView.component.ts [hunk 2]
```diff
@@ -209,9 +217,13 @@ export class AsmtResultsViewComponent extends TabChild implements IAssessmentCom
...
 			if (method === AssessmentType.InvokeAssessment) {
-				this.lastInvokedResults = result.items;
+				this.lastInvokedResult = {
+					result: result,
+					dateUpdated: Date.now(),
+					connectionInfo: this.initialConnectionInfo
+				};
 			} else {
-				this.lastInvokedResults = [];
+				this.lastInvokedResult = null;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/browser/asmtResultsView.component.ts:226]: We prefer using undefined instead of null

File: src/sql/workbench/contrib/assessment/browser/asmtView.component.ts [hunk 3]
```diff
@@ -73,7 +57,10 @@ export class AsmtViewComponent extends AngularDisposable implements OnInit {
 	}
 
 	public layout() {
-		this._asmtResultView.layout();
+		if (this._asmtResultView.layout !== undefined) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/browser/asmtView.component.ts:60]: Can shorten this to `this._asmtResultView.layout?.layout();`

File: src/sql/workbench/contrib/assessment/common/asmtActions.ts [hunk 4]
```diff
@@ -198,8 +211,8 @@ export class AsmtExportAsScriptAction extends Action {
 
 	public async run(context: IAsmtActionInfo): Promise<boolean> {
 		this._telemetryService.sendActionEvent(TelemetryView.SqlAssessment, AsmtExportAsScriptAction.ID);
-		if (context && context.component && context.component.resultItems) {
-			await this._assessmentService.generateAssessmentScript(context.ownerUri, context.component.resultItems);
+		if (context && context.component && context.component.recentResult?.result.items) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/common/asmtActions.ts:214]: Can shorten this to `if (context?.component?.recentResult?.result.items)`

File: src/sql/workbench/contrib/assessment/common/asmtActions.ts [hunk 5]
```diff
@@ -198,8 +211,8 @@ export class AsmtExportAsScriptAction extends Action {
...
+			await this._assessmentService.generateAssessmentScript(context.ownerUri, context.component.recentResult?.result.items);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/common/asmtActions.ts:215]: Since you use this twice I'd just pull it into its own var outside the if statement to keep things shorter.

File: src/sql/workbench/contrib/assessment/common/asmtActions.ts [hunk 6]
```diff
@@ -225,3 +238,60 @@ export class AsmtSamplesLinkAction extends Action {
...
+		@IClipboardService private readonly _clipboardService: IClipboardService,
+		@IAdsTelemetryService private _telemetryService: IAdsTelemetryService,
+		@INotificationService private readonly _notificationService: INotificationService
+	) {
+		super(AsmtGenerateHTMLReportAction.ID, AsmtGenerateHTMLReportAction.LABEL, AsmtGenerateHTMLReportAction.ICON);
+	}
+
+	public async run(context: IAsmtActionInfo): Promise<boolean> {
+		const fileName = this.generateReportFileName(new Date(context.component.recentResult.dateUpdated));
+		const filePath = path.join(this._environmentService.userRoamingDataHome.fsPath, 'SqlAssessmentReports', fileName);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/common/asmtActions.ts:261]: You don't want to ask them where to save it?

File: src/sql/workbench/contrib/assessment/common/asmtActions.ts [hunk 7]
```diff
@@ -225,3 +238,60 @@ export class AsmtSamplesLinkAction extends Action {
...
+		if (choice === 0) {
+			return this._openerService.open(result.resource.fsPath, { openExternal: true });
+		} else if (choice === 1) {
+			await this._clipboardService.writeText(filePath);
+			this._notificationService.info(nls.localize('asmtaction.pathcopied', 'The path has been copied to the clipboard'));
+		}
+		return true;
+	}
+
+	private generateReportFileName(resultDate): string {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/common/asmtActions.ts:291]: This can just be a standalone function - doesn't need to be a part of the class

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	INotificationHandle,
+	IPromptOptions,
+	IStatusMessageOptions,
+	NotificationsFilter,
+	NotificationMessage,
+	IPromptChoice
+} from 'vs/platform/notification/common/notification';
+
+
+export class EnvironmentServiceStub implements IEnvironmentService {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:27]: You can't use https://github.com/microsoft/azuredatastudio/blob/main/src/vs/workbench/test/browser/workbenchTestServices.ts#L31 ?1

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	logsPath: string;
+	logLevel?: string;
+	verbose: boolean;
+	isBuilt: boolean;
+	disableTelemetry: boolean;
+	serviceMachineIdResource: URI;
+
+}
+
+export class FileServiceStub implements files.IFileService {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:59]: And https://github.com/microsoft/azuredatastudio/blob/main/src/vs/workbench/test/browser/workbenchTestServices.ts#L723?

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+	watch(_resource: URI): IDisposable {
+		this.watches.push(_resource);
+
+		return null;
+	}
+	dispose(): void { }
+
+}
+
+export class DialogServiceStub implements IDialogService {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:141]: https://github.com/microsoft/azuredatastudio/blob/main/src/vs/platform/dialogs/test/common/testDialogService.ts#L9

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	show(severity: Severity, message: string, buttons: string[], options?: IDialogOptions): Promise<IShowResult> {
+		throw new Error('Method not implemented.');
+	}
+	about(): Promise<void> {
+		throw new Error('Method not implemented.');
+	}
+}
+
+export class ClipboardServiceStub implements IClipboardService {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:154]: Looks like we don't have this one yet - I'd suggest adding it to https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/test/workbenchTestServices.ts#L1 instead

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	readResources(): Promise<URI[]> {
+		return Promise.resolve(null);
+	}
+	hasResources(): Promise<boolean> {
+		return Promise.resolve(true);
+	}
+}
+
+export class NotificationServiceStub implements INotificationService {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:179]: https://github.com/microsoft/azuredatastudio/blob/main/src/vs/platform/notification/test/common/testNotificationService.ts#L9

File: src/sql/media/icons/common-icons.css [hunk 13]
```diff
@@ -233,6 +233,15 @@
 	background: url("unpin.svg") center center no-repeat;
 }
 
+.hc-black .codicon.bookreport,
```
Comment (Charles-Gagnon) [inline src/sql/media/icons/common-icons.css:236]: Just call it book. Although I'm not sure this icon is really the best option for an action that generates a report - but I'll leave that choice up to you.

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts (hunk 8 above)
Comment (vchvlad) [reply src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:27]: Yes. I'm getting the error:
`
Reading git index versions...
..\azuredatastudio\src\sql\workbench\contrib\assessment\test\common\asmtActions.test.ts: line 34, col 33, Warning - Bad layering. You are not allowed to access browser from here, allowed layers are: [common] (code-layering)
`

File: src/sql/media/icons/common-icons.css (hunk 13 above)
Comment (vchvlad) [reply src/sql/media/icons/common-icons.css:236]: "Book" is already taken by
https://github.com/microsoft/azuredatastudio/blob/6751dffbc39adff702feb2768fb2df7de26d4e45/src/sql/media/icons/common-icons.css#L558

File: src/sql/workbench/contrib/assessment/browser/asmtView.component.ts (hunk 3 above)
Comment (vchvlad) [reply src/sql/workbench/contrib/assessment/browser/asmtView.component.ts:60]: I have started getting the error: "TypeError: this._asmtResultView.layout is not a function" recently in the logs. It looks like it doesn't affect anything actually. Not sure whats the source of this issue. But to protect from this error I've added this check.

File: src/sql/media/icons/common-icons.css (hunk 13 above)
Comment (Charles-Gagnon) [reply src/sql/media/icons/common-icons.css:236]: Hrm, really wish we didn't have this mix of masked and non-masked icons...yeah ok this is fine for now.

File: src/sql/workbench/contrib/assessment/browser/asmtView.component.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/assessment/browser/asmtView.component.ts:60]: Oh whoops misread that...well that's super weird. And it's not undefined huh? Because in here we just check that the component isn't undefined which makes sense to me : https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/dashboard/browser/contents/controlHostContent.component.ts#L44

Could you try just doing that and seeing if you still see the error?

Comment (vchvlad) [reply src/sql/workbench/contrib/assessment/browser/asmtView.component.ts:60]: this._asmtResultView?.layout() leads to ""ERROR TypeError: _a.layout is not a function"

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/assessment/browser/asmtView.component.ts:60]: Huh. I don't understand how that's possible. Is this something that happens often enough to try and debug in to see what the state of the object is at that point?

angular is weird though so might just be an issue with that.

File: src/sql/workbench/contrib/assessment/common/htmlReportGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+export class HTMLReportBuilder {
+	constructor(
+		private _assessmentResult: SqlAssessmentResult,
+		private _dateUpdated: number,
+		private _connectionInfo: ConnectionManagementInfo
+	) {
+	}
+
+	public Build(): string {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/common/htmlReportGenerator.ts:20]: Typescript functions should be camelCased

File: src/sql/workbench/contrib/assessment/common/strings.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+	SERVER_VERSION: localize('asmt.serverversion', "Version:"),
+	SERVER_EDITION: localize('asmt.serveredition', "Edition:"),
+	SERVER_INSTANCENAME: localize('asmt.instancename', "Instance Name:"),
+	SERVER_OSVERSION: localize('asmt.osversion', "OS Version:"),
+	MESSAGE_COLUMN_NAME: localize('asmt.column.message', "Message"),
+	CHECKID_COLUMN_NAME: localize('asmt.column.checkId', "Check ID"),
+	LEARN_MORE_LINK: localize('asmt.learnMore', "Learn More"),
+	REPORT_TITLE: localize('asmt.sqlReportTitle', "SQL Assessment Report"),
+	RESULTS_FOR_DATABASE: localize('asmt.sqlReport.resultForDatabase', "Results for database"),
+	RESULTS_FOR_INSTANCE: localize('asmt.sqlReport.resultForInstance', "Results for instnace"),
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/assessment/common/strings.ts:22]: instance

File: src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts (hunk 8 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/assessment/test/common/vsServicesStubs.ts:27]: I'd suggest moving both the actions and tests into the browser layer then (`workbench/contrib/assessment/browser` and `workbench/contrib/assessment/test/browser`. Given that the component is in the browser anyways that shouldn't prove to be a problem and not having to duplicate all this code would be great.

File: src/sql/workbench/contrib/assessment/browser/asmtView.component.ts (hunk 3 above)
Comment (vchvlad) [reply src/sql/workbench/contrib/assessment/browser/asmtView.component.ts]: I guess this method is not actually needed at all. Previously it was necessary since the AssessmentResultView component was inside a tab of panel component. But I have moved it out of panel since we don't need tabs here.

File: src/sql/workbench/contrib/assessment/common/asmtActions.ts (hunk 6 above)
Comment (vchvlad) [reply src/sql/workbench/contrib/assessment/common/asmtActions.ts:260]: That's a really good point. I have added a SaveAs dialog to ask a user for a file location.

File: src/vs/workbench/test/browser/workbenchTestServices.ts [hunk 16]
```diff
@@ -832,7 +832,19 @@ export class TestFileService implements IFileService {
 
 	move(_source: URI, _target: URI, _overwrite?: boolean): Promise<IFileStatWithMetadata> { return Promise.resolve(null!); }
 	copy(_source: URI, _target: URI, _overwrite?: boolean): Promise<IFileStatWithMetadata> { return Promise.resolve(null!); }
-	createFile(_resource: URI, _content?: VSBuffer | VSBufferReadable, _options?: ICreateFileOptions): Promise<IFileStatWithMetadata> { return Promise.resolve(null!); }
+	createFile(resource: URI, content?: VSBuffer | VSBufferReadable, _options?: ICreateFileOptions): Promise<IFileStatWithMetadata> {
```
Comment (Charles-Gagnon) [inline src/vs/workbench/test/browser/workbenchTestServices.ts:835]: Please use TypqMoq instead to setup the value to return in your tests that need it

**Comment by Charles-Gagnon** [comment]
Thank you for adding tests!
